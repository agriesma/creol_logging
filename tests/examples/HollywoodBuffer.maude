load creol-interpreter .
mod PROGRAM is
  protecting CREOL-SIMULATOR .
  op classes : -> Configuration [ctor] .
  eq classes =
    < "Producer" : Class | Inh: noInh, Param: "buf", Att: "buf" |-> null, "i"
      |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( ( "i" @ "Producer" ) ; int(1) ) ; return ( emp )
              >,
            < "run" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "_" |-> null, Code: call( "label:0" ;
              "buf" ; "put" ; "i" ) ; get( "label:0" ; noVid ) ;
              free( "label:0" ) ; assign( "i" ; "+" ( "i" :: int(1) ) ) ;
              call( "label:1" ; "this" ; "run" ; emp ) ; free( "label:1" ) ;
              return ( emp ) >,
            < "produce" : Method | Param: noVid, Att: "x" |-> null, "_" |->
              null, Code: assign( "x" ; "i" ) ; assign( "i" ; "+" ( "i" ::
              int(1) ) ) ; return ( "x" ) >,
      Ocnt: 0 >
    
    < "Consumer" : Class | Inh: noInh, Param: "buf", Att: "buf" |-> null,
      "sum" |-> null,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( ( "sum" @ "Consumer" ) ; int(0) ) ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "label:1" |-> null,
              "label:0" |-> null, "j" |-> null, "_" |-> null,
              Code: call( "label:0" ; "buf" ; "take" ; emp ) ;
              get( "label:0" ; "j" ) ; free( "label:0" ) ;
              assign( "sum" ; "+" ( "sum" :: "j" ) ) ;
              call( "label:1" ; "this" ; "run" ; emp ) ; free( "label:1" ) ;
              return ( emp ) >,
            < "consume" : Method | Param: "x", Att: "x" |-> null, "_" |->
              null, Code: assign( "sum" ; "+" ( "sum" :: "x" ) ) ;
              return ( emp ) >,
      Ocnt: 0 >
    
    < "Buffer" : Class | Inh: noInh, Param: noVid, Att: "value" |-> null,
      "full" |-> null,
      Mtds: < "run" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "init" : Method | Param: noVid, Att: "_" |-> null,
              Code: assign( "full" ; bool(false) ) ; return ( emp ) >,
            < "put" : Method | Param: "x", Att: "x" |-> null, "_" |-> null,
              Code: await "~" ( "full" ) ; assign( "value", "full" ; "x" ::
              bool(true) ) ; return ( emp ) >,
            < "take" : Method | Param: noVid, Att: "x" |-> null, "_" |->
              null, Code: await "full" ; assign( "x", "full" ; "value" ::
              bool(false) ) ; return ( "x" ) >,
      Ocnt: 0 >
    
    < "Main" : Class | Inh: noInh, Param: noVid, Att: noSubst,
      Mtds: < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ;
              return ( emp ) >,
            < "run" : Method | Param: noVid, Att: "buf" |-> null, "prod" |->
              null, "cons" |-> null, "_" |-> null,
              Code: new( "buf" ; "Buffer" ; emp ) ;
              new( "prod" ; "Producer" ; "buf" ) ;
              new( "cons" ; "Consumer" ; "buf" ) ; return ( emp ) >,
      Ocnt: 0 > .
endm

