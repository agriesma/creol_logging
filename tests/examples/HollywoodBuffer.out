==========================================
reduce in PROGRAM : classes .
result Configuration: 
< "Buffer" : Class | Inh: noInh, Param: noVid, Att: "full" |-> null, "value"
    |-> null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("full" ;
    bool(false)) ; return(emp) >, < "put" : Method | Param: "x", Att: "_" |->
    null, "x" |-> null, Code: await "~"("full") ; assign("value", "full" ; "x"
    :: bool(true)) ; return(emp) >, < "run" : Method | Param: noVid, Att: "_"
    |-> null, Code: skip ; return(emp) >, < "take" : Method | Param: noVid,
    Att: "_" |-> null, "x" |-> null, Code: await "full" ; assign("x", "full" ;
    "value" :: bool(false)) ; return("x") >,
  Ocnt: 0 >
 
< "Consumer" : Class | Inh: noInh, Param: "buf", Att: "buf" |-> null, "sum" |->
    null, Mtds:
  < "consume" : Method | Param: "x", Att: "_" |-> null, "x" |-> null, Code:
    assign("sum" ; "+"("sum" :: "x")) ; return(emp) >, < "init" : Method |
    Param: noVid, Att: "_" |-> null, Code: assign("sum" @ "Consumer" ; int(0))
    ; return(emp) >, < "run" : Method | Param: noVid, Att: "_" |-> null, "j"
    |-> null, "label:0" |-> null, "label:1" |-> null, Code: call("label:0" ;
    "buf" ; "take" ; emp) ; get("label:0" ; "j") ; free("label:0") ; assign(
    "sum" ; "+"("sum" :: "j")) ; call("label:1" ; "this" ; "run" ; emp) ; free(
    "label:1") ; return(emp) >,
  Ocnt: 0 >
 
< "Main" : Class | Inh: noInh, Param: noVid, Att: noSubst, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: skip ; return(emp)
    >, < "run" : Method | Param: noVid, Att: "_" |-> null, "buf" |-> null,
    "cons" |-> null, "prod" |-> null, Code: new("buf" ; "Buffer" ; emp) ; new(
    "prod" ; "Producer" ; "buf") ; new("cons" ; "Consumer" ; "buf") ; return(
    emp) >,
  Ocnt: 0 >
 
< "Producer" : Class | Inh: noInh, Param: "buf", Att: "buf" |-> null, "i" |->
    null, Mtds:
  < "init" : Method | Param: noVid, Att: "_" |-> null, Code: assign("i" @
    "Producer" ; int(1)) ; return(emp) >, < "produce" : Method | Param: noVid,
    Att: "_" |-> null, "x" |-> null, Code: assign("x" ; "i") ; assign("i" ;
    "+"("i" :: int(1))) ; return("x") >, < "run" : Method | Param: noVid, Att:
    "_" |-> null, "label:0" |-> null, "label:1" |-> null, Code: call("label:0"
    ; "buf" ; "put" ; "i") ; get("label:0" ; noVid) ; free("label:0") ; assign(
    "i" ; "+"("i" :: int(1))) ; call("label:1" ; "this" ; "run" ; emp) ; free(
    "label:1") ; return(emp) >,
  Ocnt: 0 >

Bye.
